internal/venice-client-common/src/main/java/com/linkedin/venice/schema/avro/SchemaCompatibility19.java

```diff
@@ -0,0 +1,975 @@
+          }
+          case BYTES: {
+            return (writer.getType() == Type.STRING)
+                ? result
+                : result.mergedWith(typeMismatch(reader, writer, location));
+          }
+          case STRING: {
+            return (writer.getType() == Type.BYTES)
+                ? result
+                : result.mergedWith(typeMismatch(reader, writer, location));
+          }
+
+          case ARRAY:
+            return result.mergedWith(typeMismatch(reader, writer, location));
+          case MAP:
+            return result.mergedWith(typeMismatch(reader, writer, location));
+          case FIXED:
+            return result.mergedWith(typeMismatch(reader, writer, location));
+          case ENUM:
+            return result.mergedWith(typeMismatch(reader, writer, location));
+          case RECORD:
+            return result.mergedWith(typeMismatch(reader, writer, location));
+          case UNION: {
+            for (final Schema readerBranch: reader.getTypes()) {
+              SchemaCompatibilityResult compatibility = getCompatibility(readerBranch, writer);
+              if (compatibility.getCompatibility() == SchemaCompatibilityType.COMPATIBLE) {
+                return result;
+              }
+            }
+            // No branch in the reader union has been found compatible with the writer
+            // schema:
+            String message = String.format("reader union lacking writer type: %s", writer.getType());
+            return result.mergedWith(
+                SchemaCompatibilityResult.incompatible(
+                    SchemaIncompatibilityType.MISSING_UNION_BRANCH,
+                    reader,
+                    writer,
+                    message,
+                    asList(location)));
+          }
+
+          default: {
+            throw new AvroRuntimeException("Unknown schema type: " + reader.getType());
+          }
+        }
+      }
+    }
+
+    private SchemaCompatibilityResult checkReaderWriterRecordFields(
+        final Schema reader,
+        final Schema writer,
+        final Deque<String> location) {
+      SchemaCompatibilityResult result = SchemaCompatibilityResult.compatible();
+      location.addFirst("fields");
+      // Check that each field in the reader record can be populated from the writer
+      // record:
+      for (final Field readerField: reader.getFields()) {
+        location.addFirst(Integer.toString(readerField.pos()));
+        final Field writerField = lookupWriterField(writer, readerField);
+        if (writerField == null) {
+          // Reader field does not correspond to any field in the writer record schema, so
+          // the
+          // reader field must have a default value.
+          if (!readerField.hasDefaultValue()) {
+            // reader field has no default value. Check for the enum default value
+            if (readerField.schema().getType() == Type.ENUM && readerField.schema().getEnumDefault() != null) {
```

Comment:
ditto: `AvroCompatibilityHelper.getEnumDefault(readerField.schema()) != null`

Reply:
Based on the internal discussion, we will keep it as it is for now.